### PR TITLE
feat: pyarrow automated inference of format `%Y%m%d%H%M%S` 

### LINKED ISSUE #1282: [Enh]: Increase support for datetime format in pyarrow automatic inference
### Please describe the purpose of the new feature or describe the problem to solve.

As discussed in #1151 [comment](https://github.com/narwhals-dev/narwhals/pull/1195#issuecomment-2443859109), let's try to expand the available formats that we support with automatic inference in pyarrow `str.to_datetime`. Non exhaustive list:

- [x] `"%Y-%m-%dT%H:%M"`
- [ ] `"%Y%m%d%H%M%S"`

### PR DESCRIPTION
**Pull request by raisadz**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #1282

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added 
- [ ] Documented the changes

### CODE REVIEW

**Review by FBruzzesi — COMMENTED** [review]
Thanks @raisadz !

I didn't know that polars does not detect this format automatically. Since that's the case, I feel a bit hesitant to add it honestly.
Maybe let's sleep a bit on this and wait for Marco weighting on it

File: tests/expr_and_series/str/to_datetime_test.py
[the remark this answers is not in the document]
Comment (raisadz) [reply tests/expr_and_series/str/to_datetime_test.py:1]: Thank you for the suggestion, @EdAbati! I agree that it is cleaner to add the new case to the existing tests, I addressed this.

File: narwhals/_arrow/utils.py [hunk 1]
```diff
@@ -350,28 +350,28 @@ def convert_str_slice_to_int_slice(
...
 
 # Regex for date, time, separator and timezone components
-DATE_RE = r"(?P<date>\d{1,4}[-/.]\d{1,2}[-/.]\d{1,4})"
+DATE_RE = r"(?P<date>\d{1,4}[-/.]\d{1,2}[-/.]\d{1,4}|\d{8})"
 SEP_RE = r"(?P<sep>\s|T)"
-TIME_RE = r"(?P<time>\d{2}:\d{2}(?::\d{2})?)"  # \s*(?P<period>[AP]M)?)?
+TIME_RE = r"(?P<time>\d{2}:\d{2}(?::\d{2})?|\d{6}?)"  # \s*(?P<period>[AP]M)?)?
 HMS_RE = r"^(?P<hms>\d{2}:\d{2}:\d{2})$"
 HM_RE = r"^(?P<hm>\d{2}:\d{2})$"
+HMS_RE_NO_SEP = r"^(?P<hmsns>\d{6})$"
```
Comment (MarcoGorelli) [inline narwhals/_arrow/utils.py:358]: "ns" reads to me as "nanoseconds" here - maybe we can just write `hms_no_sep`?

**Review by MarcoGorelli — APPROVED** [review]
thanks for your PR (and @EdAbati + @FBruzzesi for reviews)

just got one really minor comment

regarding Polars, the datetime inference there is just based on trying out a bunch of hardcoded formats, and is probably ripe for being rewritten anyway

so i think we might as well go ahead with this

https://github.com/pola-rs/polars/blob/750b3712fa36046a9a2e35cdaddf5f767f3febab/crates/polars-time/src/chunkedarray/string/patterns.rs#L84-L151

**Pull request merged**